libfeed/feed-store-provider.c

```diff
@@ -0,0 +1,1239 @@
+                         GError      **error)
+{
+  gint64 ret = 0;
+
+  g_return_val_if_fail (out != NULL, FALSE);
+
+  /* Clear errno first before calling g_ascii_stroll */
+  errno = 0;
+
+  ret = g_ascii_strtoll (str, NULL, base);
+
+  if (errno != 0)
+    {
+      g_set_error (error,
+                   G_IO_ERROR,
+                   g_io_error_from_errno (errno),
+                   "Parsing of integer failed with %s", strerror (errno));
+      return FALSE;
+    }
+
+  if (ret < min || ret > max)
+    {
+      g_set_error (error,
+                   G_IO_ERROR,
+                   G_IO_ERROR_FAILED,
+                   "Parsing of integer failed, %lli is not "
+                   "in the range of %lli - %lli",
+                   ret,
+                   min,
+                   max);
+      return FALSE;
+    }
+
+  *out = ret;
+  return TRUE;
+}
+
+static gint64
+parse_base10_int64_with_limits (const gchar  *str,
+                                gint64        min,
+                                gint64        max,
+                                gint64       *out,
+                                GError      **error)
+{
+  g_return_val_if_fail (out != NULL, FALSE);
+
+  return parse_int64_with_limits (str, 10, min, max, out, error);
+}
+
+static char *
+parse_duration (const gchar  *duration,
+                GError      **error)
+{
+  gint64 total_seconds = 0;
+  gint64 hours = 0;
+  gint64 minutes = 0;
+  gint64 seconds = 0;
+
+  if (!parse_base10_int64_with_limits (duration,
+                                       G_MININT64,
+                                       G_MAXINT64,
+                                       &total_seconds,
+                                       error))
+    return NULL;
+
+  hours = floor (total_seconds / 3600);
+  minutes = ((gint64) floor (total_seconds / 60)) % 60;
+  seconds = floor (((gint64) total_seconds) % 60);
+
+  if (hours > 0)
+    return g_strdup_printf ("%lli:%02lli:%02lli", hours, minutes, seconds);
+
+  return g_strdup_printf ("%lli:%02lli", minutes, seconds);
+}
+
+typedef struct _VideoCardsFromShardsAndItemsData
+{
+  EosDiscoveryFeedKnowledgeAppProxy                *ka_proxy;
+  EosDiscoveryFeedCardStoreType                     type;
+} VideoCardsFromShardsAndItemsData;
+
+static VideoCardsFromShardsAndItemsData *
+video_cards_from_shards_and_items_data_new (EosDiscoveryFeedKnowledgeAppProxy *ka_proxy,
+                                            EosDiscoveryFeedCardStoreType      type)
+{
+  VideoCardsFromShardsAndItemsData *data = g_new0 (VideoCardsFromShardsAndItemsData, 1);
+
+  data->ka_proxy = g_object_ref (ka_proxy);
+  data->type = type;
+
+  return data;
+}
+
+static void
+video_cards_from_shards_and_items_data_free (VideoCardsFromShardsAndItemsData *data)
+{
+  g_clear_object (&data->ka_proxy);
+
+  g_free (data);
+}
+
+G_DEFINE_AUTOPTR_CLEANUP_FUNC (VideoCardsFromShardsAndItemsData,
+                               video_cards_from_shards_and_items_data_free);
```

Comment:
This can all be removed in favour of just hardcoding `EOS_DISCOVERY_FEED_CARD_STORE_TYPE_VIDEO_CARD` below

Reply:
Ah, we'll still need to pass the KA proxy as well, but maybe we can just pass that on its own as opposed to using a closure.